benchmark/src/main/java/com/linecorp/decaton/benchmark/ForkingExecution.java

```diff
@@ -50,67 +51,67 @@
 
     public static void main(String[] args) throws IOException, InterruptedException {
         Config config = mapper.readValue(args[0], Config.class);
+        RemoteCallback remoteCallback = BenchmarkRmi.lookup(Integer.parseInt(args[1]));
+
         InProcessExecution execution = new InProcessExecution();
-        BenchmarkResult result = execution.execute(config, System.out::println);
-        mapper.writeValue(System.out, result);
+        BenchmarkResult result = execution.execute(config, stage -> {
+            try {
+                remoteCallback.onStage(stage);
+            } catch (RemoteException e) {
+                log.error("Failed to invoke remote callback", e);
+                throw new RuntimeException(e);
+            }
+        });
+
+        remoteCallback.onResult(mapper.writeValueAsString(result));
     }
 
     @Override
     public BenchmarkResult execute(Config config, Consumer<Stage> stageCallback) {
         List<String> cmd = new ArrayList<>();
-        cmd.add(javaBin().toString());
-        cmd.add("-server");
-        cmd.addAll(jvmFlags());
-        cmd.add("-cp"); cmd.add(currentClasspath());
-        cmd.add(ForkingExecution.class.getName());
-        cmd.add(serializeConfig(config));
 
+        final BenchmarkResult result;
         final Process process;
-        try {
+        try (BenchmarkRmi rmi = new BenchmarkRmi()) {
+            CompletableFuture<String> resultFuture = rmi.start();
+
+            cmd.add(javaBin().toString());
+            cmd.add("-server");
+            cmd.addAll(jvmFlags());
+            cmd.add("-cp"); cmd.add(currentClasspath());
+            cmd.add(ForkingExecution.class.getName());
+            cmd.add(serializeConfig(config));
+            cmd.add(String.valueOf(rmi.port()));
+
             log.debug("Forking child process for run with: {}", cmd);
             process = new ProcessBuilder()
                     .command(cmd)
                     .redirectError(Redirect.INHERIT)
+                    .redirectOutput(Redirect.INHERIT)
                     .start();
+
+            Stage stage;
+            while ((stage = rmi.pollStage()) != Stage.FINISH) {
```

Comment:
any points to put a queue in between for this? As we anyway forward the event to the another callback, would it be more natural to give the callback itself to Rmi from the beginning?

Reply:
For this point: https://github.com/line/decaton/pull/68/files#diff-308fef37ee55388e364ad04c5992d6e2R40
wdyt?

Reply:
As for separating RMI thread with processing threads, it seems reasonable.